examples/screenshot_and_export_figure.py

```diff
@@ -0,0 +1,104 @@
+"""
+Comparison of Screenshot and Figure Export
+==========================================
+
+Display multiple layer types, add scale bar, and take a screenshot or export a
+figure from a 'light' canvas. Then switch to a 'dark' canvas and display the
+screenshot and figure. Compare the limits of each export method. The screenshot
+will include the entire canvas, and results in some layers being clipped
+if it extends outside the canvas. In comparison, the `export_figure`
+will only include the extent of the layers and any other elements overlayed
+on the canvas, such as the scale bar. Exported figures also move the scale bar
+to within the margins of the canvas.
+
+Currently, 'export_figure` does not support the 3D view, but screenshot does.
+
+.. tags:: visualization-advanced
+"""
+
+import numpy as np
+from skimage import data
+
+import napari
+
+##############################################################################
+# Create a napari viewer with multiple layer types and add a scale bar.
+# One of the polygon shapes exists outside the image extent, which is
+# useful in displaying how figure export handles the extent of all layers.
+
+viewer = napari.Viewer()
+
+# add a 2D image layer
+img_layer = viewer.add_image(data.camera(), name='photographer')
+img_layer.colormap = 'gray'
+
+# polygon within image extent
+layer_within = viewer.add_shapes(
+    np.array([[11, 13], [111, 113], [22, 246]]),
+    shape_type='polygon',
+    face_color='coral',
+    name='shapes_within',
+)
+
+# add a polygon shape layer
+layer_outside = viewer.add_shapes(
+    np.array([[572, 222], [305, 292], [577, 440]]),
+    shape_type='polygon',
+    face_color='royalblue',
+    name='shapes_outside',
+)
+
+# add scale_bar with background box
+viewer.scale_bar.visible = True
+viewer.scale_bar.box = True
```

Comment:
I don't think so but I honestly don't know - maybe @lucyleeow does?

Reply:
I don't think there is a time out. I had a quick dig into Sphinx Gallery and we just iterate through all scrapers, for each code block to get the output, see [here](https://github.com/sphinx-gallery/sphinx-gallery/blob/80dbfe20c4962fc6895d15f95b43a27de156e7f6/sphinx_gallery/scrapers.py#L389).

So for each code block, the napari scraper code is run, which means the viewer will be closed and then the rst of the image is returned. Is this error occurring when running `Viewer.close_all()`?

Reply:
No. The `Viewer.close_all()` is called in the middle of example code, and in followup lines, that are tried to access to deleted structures, causes exception.

Reply:
Just in case it was missed, dropping the code block syntax did fix the issue earlier.
but obviously we'd like to keep it.

Reply:
Sorry, can you expand?

Reply:
What did you mean by code blocks?
Did scrapper is called on every line with `#######################################################################` ?

Reply:
This example isn't handled like a simple script because it uses the `########` formatting to make code blocks. If you remove that to make it handled a regular script, then there are no issues with the docs locally.

Reply:
Ok. So we know what is the source of the problem. @psobolewskiPhD could you provide detailed description of possible solutions to @TimMonko?

Reply:
@Czaki yes, line of `#`s or `# %%` splits example into blocks. Text block immediately follows a 'splitter', which is followed by a code block when you stop the comments (`#` at start of line).
(details here: https://sphinx-gallery.github.io/stable/syntax.html#embed-rest-in-your-example-python-files)

Reply:
Yea, we have other examples that use this method with no problems. What's unique about this? Seems vanilla.

Reply:
I linked it above, but here it is again:
https://napari.org/stable/gallery/vortex.html
https://github.com/napari/napari/blob/main/examples/vortex.py

Reply:
As you may see, the viewer is created and used only in the last cell.

Reply:
Been following this discussion. It's been enlightening to read everyone's thoughts. 

Here is [an example in the gallery that has multiple calls to the viewer](https://napari.org/stable/gallery/dynamic-projections-dask.html), but I think if I'm understanding correctly the viewer is recreated in each code block. 

A difference to this pull's example is that the same viewer is used in each code block. If @Czaki is saying that the viewer gets deconstructed at the end of each block, then this example would need to be modified. 

One solution would be to not use the reST blocks, but I think another solution would be if there was a way to persist the viewer through blocks. 

Is the solution actually just 
```
if __name__ == '__main__':
    napari.run()
```
Truth be told, I've never really understood this syntax

Reply:
> Is the solution actually just
> 
> ```
> if __name__ == '__main__':
>     napari.run()
> ```
> Truth be told, I've never really understood this syntax

This syntax is to start Qt event loop only if the example is executed as script, and do not start it when it is executed during gallery build 


> One solution would be to not use the reST blocks, but I think another solution would be if there was a way to persist the viewer through blocks. Ideally, the code blocks would allow the viewer to persist through (like would be the case with some like a jupyter notebook).

The [reset_modules](https://sphinx-gallery.github.io/stable/advanced.html#resetting-before-each-example) could be used for cleaning, but then I do not know how to found which viewers should be used for taking screenshot.